Allow bench_refs() and bench_local_refs() to capture input lifetimes

neocturne opened the pull request:
This allows benchmarks using bench_refs() and bench_local_refs() to capture the input lifetimes, as was already suggested by TODO comments in the code.

By naming the input lifetime, the output type `O` is allowed to capture this lifetime. The `I: 'i` condition for `&'i mut I` must be made explicit now with named `'i`, or the code won't compile.

First commit adds tests that don't compile, second commit makes them compile.

neocturne committed:
bench/tests: Add non-working capturing bench_refs() tests

Prepare tests that currently fail to compile, showing that bench outputs
can't capture the input lifetime.

neocturne committed:
bench: allow bench_refs() and bench_local_refs() to capture input lifetimes

By naming the input lifetime, the output type `O` is allowed to capture
this lifetime. The `I: 'i` condition for `&'i mut I` must be made
explicit now with named `'i`, or the code won't compile.

This makes all Capturing tests added in the previous commit compile.

nvzqz reviewed (commented):
Thanks for this change!

Some suggestions regarding testing:
- I like the internal tests with `Capturing` and I think it would make sense to have it capture `&'a mut ()` to restrict it based on the parameter.
- Let's include an example benchmark using this feature.

neocturne commented:
Any suggestions for a good example? In my project I'm using it for benchmarking a parser that creates a token stream/tree which references the input, but that seems too complex for an example (and it uses a separate parser crate). I guess I could do something simple like collecting the result of `str::matches()` in a `Vec` that gets returned?

neocturne committed:
examples: string: add "matches" example that captures the input lifetime

Add a simple example that shows that bench_refs/bench_local_refs input
lifetimes may be captured now.

neocturne commented:
Any further comments?

neocturne commented:
@nvzqz Is there anything else you'd like me to change in this PR?

nvzqz commented:
Hey @neocturne, sorry for the late response, and thanks again for the contribution. Since I'm not sure whether adding a lifetime here is a breaking change, and I already have some breaking changes planned, this will land in `0.2` soon.

neocturne commented:
I saw that there was a merge conflict - rebased onto main now.

nvzqz commented:
When coming back to this, I realized that this approach allows the input reference to escape the benchmarking loop, so this approach would be unsound.

```rs
let mut escaped_ref = &mut String::new();

bencher
    .with_inputs(|| "hi".to_owned())
    .bench_local_refs(|input: &mut String| escaped_ref = input);

println!("{escaped_ref}");
```

Given this is such a niche use case, I would rather people transmute the lifetime to `'static`, as opposed to making this approach available as a secondary `unsafe` function.

Thanks for the contribution, though!
